File: kotlin/day1/day1.kt
import java.io.File
import java.io.FileReader
import kotlin.math.abs

fun pt1(lines: List<String>): Int {
    val A = mutableListOf<Int>()
    val B = mutableListOf<Int>()
  for (line in lines) {
    val (a, b) = line.split("   ")
    A.add(a.toInt())
    B.add(b.toInt())

  }
  val ans = A.sorted().zip(B.sorted()) { a, b -> 
    abs(a - b)
  }.sum()
  println(ans)

    return 0
}

fun pt2(lines: List<String>): Int {
    val A = mutableListOf<Int>()
    val B = mutableMapOf<Int, Int>()
  for (line in lines) {
    val (a, b) = line.split("   ")
    A.add(a.toInt())
    if (b.toInt() in B.keys) {
      B[b.toInt()] = B[b.toInt()]!! + 1
    } else {
      B[b.toInt()] = 1
    }

  }
  val ans = A.map { it * B.getOrDefault(it, 0) }.sum()
  println(ans)
  return ans

}

fun main() {
    val input = FileReader(File("../../inputs/day1.txt")).readLines()
    println(pt1(input))
    println(pt2(input))
}

File: python/day01/day1.py
import re
import heapq
from collections import deque, defaultdict, Counter


input = open("../../inputs/day1.txt").readlines()


def part1(lines):
    a = []
    b = []
    for line in lines:
        x, y = line.split()
        a.append(int(x.strip()))
        b.append(int(y.strip()))
    a.sort()
    b.sort()
    ans = 0
    for p, q in zip(a, b):
        t = abs(p - q)
        ans += t

    return ans


def part2(lines):
    a = []
    b = []
    for line in lines:
        x, y = line.split()
        a.append(int(x.strip()))
        b.append(int(y.strip()))
    freq = Counter(b)
    ans = 0
    for p in a:
        if p in freq:
            f = freq[p]
            ans += p * f

    return ans


print(part1(input))
print(part2(input))
